test/rgb_matrix/animation_test.exs

```diff
@@ -1,12 +1,190 @@
 defmodule RGBMatrix.AnimationTest do
   use ExUnit.Case
 
+  alias Layout.LED
   alias RGBMatrix.Animation
+  alias RGBMatrix.Animation.Config.FieldType.{Integer, Option}
+
+  setup_all [
+    :create_leds,
+    :create_frame,
+    :create_animation_module,
+    :create_config,
+    :create_animation,
+    :create_schema
+  ]
 
   test "can get an animation type's human-readable name" do
-    defmodule MockAnimation.HueWave do
+    assert Animation.type_name(MockHueWave) == "Mock Hue Wave"
+    assert Animation.type_name(MockAnimations.Mock.HueWave) == "Hue Wave"
+    assert Animation.type_name(Mock.Animations.Hue.Wave) == "Wave"
+  end
+
+  test "can create a new animation", %{
+    animation: animation,
+    animation_module: animation_module,
+    leds: leds
+  } do
+    assert ^animation = Animation.new(animation_module, leds)
+  end
+
+  test "rendering returns a delay, frame, and the maybe-updated animation", %{
+    animation: animation,
+    config: config,
+    frame: frame
+  } do
+    assert {1000, ^frame, animation} = Animation.render(animation)
+    assert animation.state == [:render, :new]
+    assert animation.config == config
+  end
+
+  test "interaction returns a delay and the maybe-updated animation", %{
+    animation: animation,
+    config: config,
+    leds: leds
+  } do
+    led = hd(leds)
+
+    assert {0, animation} = Animation.interact(animation, led)
+    assert animation.state == [{:interact, led}, :new]
+    assert animation.config == config
+  end
+
+  test "can get the config and schema", %{
+    animation: animation,
+    config: config,
+    schema: schema
+  } do
+    assert Animation.get_config(animation) == {config, schema}
+  end
+
+  test "can update the config of the provided animation", %{
+    animation: animation,
+    config: config
+  } do
+    params = %{"test_integer" => "5", test_option: :b}
+    updated_config = %{config | test_option: :b, test_integer: 5}
+
+    assert animation = %Animation{} = Animation.update_config(animation, params)
+    refute animation.config == config
+    assert animation.config == updated_config
```

Comment:
ℹ️ (optional / question)

Doesn't the nature of `assert animation.config == updated_config` mean that if it passes the `refute` will always fail, and vice versa? In other words, only one of these tests is necessary (preferably the assertion verifying the intended state -- `assert`).

Reply:
Hmm... I see your point, yeah. I suppose an incorrectly-updated config should be caught by the Config tests, not the Animation tests.